File: network/src/main/java/pl/malopolska/smoksmog/utils/StationUtils.java
package pl.malopolska.smoksmog.utils;


import java.util.Collection;
import java.util.List;

import pl.malopolska.smoksmog.model.Station;
import rx.Observable;
import rx.functions.Func1;

public class StationUtils {

    private StationUtils() {
        throw new IllegalAccessError( "Utils class" );
    }

    /**
     * Converts station list to list of station ids
     *
     * @param stationList
     * @return
     */
    public static List<Long> convertStationsToIdsList( List<Station> stationList ) {
        return Observable.from( stationList )
                .map( new Func1<Station, Long>() {
                    @Override
                    public Long call( Station station ) {
                        return station.getId();
                    }
                } )
                .toList().toBlocking().first();
    }

    /**
     *
     *
     * @param stations
     * @return
     */
    public static long[] convertStationsToIdsArray( Collection<Station> stations ){
        long[] result = new long[stations.size()];
        int i = 0;
        for( Station station : stations ){
            result[i] = station.getId();
            i++;
        }
        return result;
    }

    /**
     * @param stations
     * @param longitude
     * @param latitude
     * @return
     */
    public static Station findClosest( Collection<Station> stations, double latitude, double longitude ) {

        Station closestStation = null;
        double distance = Double.MAX_VALUE;

        for ( Station station : stations ) {

            double calculatedDistance = distanceInRadians(
                    station.getLatitude(), station.getLongitude(),
                    latitude, longitude );

            if ( calculatedDistance < distance ) {
                closestStation = station;
                distance = calculatedDistance;
            }
        }

        return closestStation;
    }

    /**
     * @param latitude1
     * @param longitude1
     * @param latitude2
     * @param longitude2
     * @return
     */
    private static double distanceInRadians( double latitude1, double longitude1, double latitude2, double longitude2 ) {

        double x1 = Math.toRadians( latitude1 );
        double y1 = Math.toRadians( longitude1 );
        double x2 = Math.toRadians( latitude2 );
        double y2 = Math.toRadians( longitude2 );

        // great circle distance in radians
        return Math.acos( Math.sin( x1 ) * Math.sin( x2 ) + Math.cos( x1 ) * Math.cos( x2 ) * Math.cos( y1 - y2 ) );
    }
}

File: network/src/main/kotlin/pl/malopolska/smoksmog/utils/StationUtils.kt
package pl.malopolska.smoksmog.utils


import pl.malopolska.smoksmog.model.Station
import rx.Observable

class StationUtils private constructor() {

    init {
        throw IllegalAccessError("Utils class")
    }

    companion object {

        fun convertStationsToIdsList(stationList: List<Station>): List<Long> {
            return Observable.from(stationList).map { station -> station.id }.toList().toBlocking().first()
        }

        fun convertStationsToIdsArray(stations: Collection<Station>): LongArray {
            val result = LongArray(stations.size)
            var i = 0
            for ((id) in stations) {
                result[i] = id
                i++
            }
            return result
        }

        fun findClosest(stations: Collection<Station>, latitude: Double, longitude: Double): Station {

            var closestStation: Station = stations.first()
            var distance = java.lang.Double.MAX_VALUE

            for (station in stations) {

                val calculatedDistance = distanceInRadians(
                        station.latitude.toDouble(), station.longitude.toDouble(),
                        latitude, longitude)

                if (calculatedDistance < distance) {
                    closestStation = station
                    distance = calculatedDistance
                }
            }

            return closestStation
        }

        private fun distanceInRadians(latitude1: Double, longitude1: Double, latitude2: Double, longitude2: Double): Double {

            val x1 = Math.toRadians(latitude1)
            val y1 = Math.toRadians(longitude1)
            val x2 = Math.toRadians(latitude2)
            val y2 = Math.toRadians(longitude2)

            // great circle distance in radians
            return Math.acos(Math.sin(x1) * Math.sin(x2) + Math.cos(x1) * Math.cos(x2) * Math.cos(y1 - y2))
        }
    }
}
